packages/workspace/src/browser/workspace-service.ts

```diff
@@ -88,78 +83,103 @@ export class WorkspaceService implements FrontendApplicationContribution {
         return this.onWorkspaceChangeEmitter.event;
     }
 
+    protected async onWorkspaceFileChanged(newWorkspaceFile?: FileStat): Promise<void> {
+        if (newWorkspaceFile) {
+            if (this._workspaceFile && this._workspaceFile.uri !== newWorkspaceFile.uri) {
+                this.stopWatch(this._workspaceFile.uri);
+            }
+            this._workspaceFile = newWorkspaceFile;
+        }
+        this.updateTitle();
+        if (this._workspaceFile) {
+            await this.doWatch(this._workspaceFile);
+            this.watcher.onFilesChanged(event => {
+                if (FileChangeEvent.isAffected(event, new URI(this._workspaceFile!.uri).withScheme('file'))) {
+                    this.updateWorkspace();
+                }
+            });
+        }
+    }
+
     protected async updateWorkspace(): Promise<void> {
-        await this.ensureWorkspaceInitialized();
-        await this.updateRoots();
+        const rootUri = await this.server.getMostRecentlyUsedWorkspace();
```

Comment:
Why do we reread it each time whenever a workspace file changed? Should not we read it only once as before just to get a workspace file and then call with it `setWorkspaceFile` first time? `setWorkspaceFile` also should be responsible for watching a workspace file not `updateWorkspace`.

Reply:
I don't understand it:
- first, `updateWorkspace` should not change `_worksapceFile` or its watcher, it can be only done in `setWorkspaceFile`. `updateWorkspace` should operate like a workspace file is already up to date and only manage roots. Because of it reading of most recently used workspace should be moved out from here to `init` method:
```ts
    @postConstruct()
    protected async init(): Promise<void> {
        const workspaceUri = await this.server.getMostRecentlyUsedWorkspace();
        const workspaceFile = await this.toFileStat(workspaceUri);
        await this.setWorkspaceFile(workspaceFile);
        this.preferences.onPreferenceChanged(event => {
            if (event.preferenceName === 'workspace.supportMultiRootWorkspace') {
                this.updateWorkspace();
            }
        });
    }
```
- second, all tabs read the same workspace file on init, if it gets changed they get updated, since they listen to in. if one of tabs switches to another workspace file, why should we switch others?